### PR TITLE
@ember routing Specify that the returned Transition resolves to a Route

### PR DESCRIPTION
**Pull request by mongoose700**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: Ultimately the definitions for [how the router gets set up in Ember](https://github.com/emberjs/ember.js/blob/v3.25.1/packages/%40ember/-internals/routing/lib/system/router.ts#L129), and [how it uses that specified type](https://github.com/tildeio/router.js/blob/master/lib/router/router.ts#L298). I haven't been able to find any documentation for it.
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header. (It does not)

It may be that the intent is not to reveal that the transition resolves to a route, in which case it would probably be better to make it explicitly a `Transition<void>` instead of leaving it as a `Transition<any>`.

### CODE REVIEW

File: types/ember__routing/-private/transition.d.ts [hunk 1]
```diff
@@ -42,7 +42,7 @@ export default interface Transition<T = any> extends Partial<Promise<T>> {
     catch<TResult = never>(
         onRejected?: (reason: any) => TResult | PromiseLike<TResult>,
         label?: string,
-    ): Promise<TResult>;
+    ): Promise<TResult | T>;
```
Comment (mongoose700) [inline types/ember__routing/-private/transition.d.ts:45]: This is a separate bugfix I noticed that brings it more inline with `Promise` and matches the actual behavior.

**Review by chriskrycho — CHANGES_REQUESTED** [review]
Sorry, only just saw this. Thanks for submitting it!

As you suggest, what a `Transition` resolves to is intentionally not specified. The better definition here is probably to change it to `Transition<unknown>`, which correctly captures the semantics: there *is* an object resolved there, but it's basically opaque to end users. (While `void` would prevent narrowing, it is also incorrect, which is what makes it less preferable here.)

This highlights an important related point, though: As we think about these types going forward as part of the RFC and design process for official TS support in Ember, we should take care to make sure these aren't under-specified! We may want to introduce explicit opaque types which cannot further be narrowed and which have the correct semantics, for example!

In any case, if you can update it to return `Transition<unknown>`, that would be a good iterative step here!

**Comment by mongoose700** [comment]
I modified the `Transition` itself to default to `unknown` instead of `any`, instead of making the change only in the route transitions.

**Review by chriskrycho — APPROVED** [review]
Thank you!

**Comment by chriskrycho** [comment]
Ready to merge

**Pull request merged**